src/rules/no-reactive-literals.ts

```diff
@@ -0,0 +1,69 @@
+import type { TSESTree } from "@typescript-eslint/types"
+import { createRule } from "../utils"
+
+const labeledStatementBase = `SvelteReactiveStatement > ExpressionStatement > AssignmentExpression`
+
+export default createRule("no-reactive-literals", {
+  meta: {
+    docs: {
+      description: "Don't assign literal values in reactive statements",
+      category: "Stylistic Issues",
+      recommended: false,
+      conflictWithPrettier: false,
+    },
+    fixable: "code",
+    schema: [],
+    messages: {
+      noReactiveLiterals: `Do not assign literal values inside reactive statements unless absolutely necessary.`,
+    },
+    type: "suggestion",
+  },
+  create(context) {
+    /**
+     * Reusable method for multiple types of nodes that should warn
+     *
+     * @param node TSESTree.AssignmentExpression the node that was found
+     * @returns void
+     */
+    function warn(node: TSESTree.AssignmentExpression) {
+      // Move upwards to include the entire reactive statement
+      const parent = node.parent?.parent
+
+      if (!parent) {
+        return false
+      }
+
+      const source = context.getSourceCode()
+
+      return context.report({
+        node: parent,
+        loc: parent.loc,
+        messageId: "noReactiveLiterals",
+
+        fix(fixer) {
+          return [
+            // Insert "let" + whatever was in there
+            fixer.insertTextBefore(parent, `let ${source.getText(node)}`),
+
+            // Remove the original reactive statement
+            fixer.remove(parent),
+          ]
+        },
+      })
+    }
+
+    return {
+      // $: foo = "foo";
+      // $: foo = 1;
+      [`${labeledStatementBase}[right.type="Literal"]`]: warn,
+
+      // $: foo = [];
+      [`${labeledStatementBase}[right.type="ArrayExpression"][right.elements.length=0]`]:
+        warn,
+
+      // $: foo = {};
+      [`${labeledStatementBase}[right.type="ObjectExpression"][right.properties.length=0]`]:
+        warn,
```

Comment:
I think you can make it simpler by using `:matches()`.

```suggestion
      [`SvelteReactiveStatement > ExpressionStatement > AssignmentExpression:matches(${[
        // $: foo = "foo";
        // $: foo = 1;
        `[right.type="Literal"]`,
        // $: foo = [];
        `[right.type="ArrayExpression"][right.elements.length=0]`,
        // $: foo = {};
        `[right.type="ObjectExpression"][right.properties.length=0]`,
      ].join(",")})`]: warn,
```

https://github.com/estools/esquery#:~:text=matches%2Dany%3A-,%3Amatches,-(%5Battr%5D%20%3E%20%3Afirst%2Dchild